From 12449f615e4240241d6ec16b0bb7b404354d0eb6 Mon Sep 17 00:00:00 2001
From: Yaroslav Halchenko <debian@onerussian.com>
Date: Wed, 21 Oct 2015 11:31:18 -0500
Subject: [PATCH] BF: allow to "compare" SmartVersion(None)

Should fix test_generate_testing_fmri_dataset which relies on comparison
of versions of ipython which might be None
---
 mvpa2/misc/support.py       | 11 +++++++++++
 mvpa2/tests/test_support.py |  4 ++++
 2 files changed, 15 insertions(+)

diff --git a/mvpa2/misc/support.py b/mvpa2/misc/support.py
index fdf885e..af02ecb 100644
--- a/mvpa2/misc/support.py
+++ b/mvpa2/misc/support.py
@@ -415,6 +415,17 @@ class SmartVersion(Version):
 
         # Do ad-hoc comparison of strings
         i = 0
+
+        # if any of the versions was not parsed (e.g. if None was provided),
+        # comparison can't be performed really unless both have no version
+        # assigned
+        if (not hasattr(self, 'version')) and (not hasattr(other, 'version')):
+            return 0
+
+        for v in (self, other):
+            if not (hasattr(v, 'version')):
+                raise ValueError('%s has no version information' % v)
+
         s, o = self.version, other.version
         regex_prerelease = re.compile('~|-?dev|-?rc|-?svn|-?pre|-?beta|-?alpha', re.I)
         for i in xrange(max(len(s), len(o))):
diff --git a/mvpa2/tests/test_support.py b/mvpa2/tests/test_support.py
index 3ee5dbe..2c72099 100644
--- a/mvpa2/tests/test_support.py
+++ b/mvpa2/tests/test_support.py
@@ -217,6 +217,10 @@ class SupportFxTests(unittest.TestCase):
         """
         SV = SmartVersion
 
+        self.assertEqual(SmartVersion(None), SmartVersion(None))
+        self.assertRaises(ValueError, SmartVersion(None).__cmp__, SmartVersion('0'))
+        self.assertRaises(ValueError, SmartVersion('0').__cmp__, SmartVersion(None))
+
         for v1_, v2_ in (
             ('0.0.1', '0.0.2'),
             ('0.0.1', '0.1'),
-- 
2.5.1

